### PR TITLE
[Local Testing] SNOW-904259 Support DataFrame.DropDuplicates

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Review by sfc-gh-sfan — COMMENTED** [review]
To be frank my biggest confusion is: We are enabling `DataFrame#DropDuplicates`, but we don't have any change in `drop_duplicates` function 🤔

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -3434,20 +3434,25 @@ def cache_result(
...
-            create_temp_table,
-            _statement_params=create_or_update_statement_params_with_query_tag(
-                statement_params or self._statement_params,
-                self._session.query_tag,
-                SKIP_LEVELS_TWO,
-            ),
-        )
+        from snowflake.snowpark.mock.connection import MockServerConnection
+
+        if isinstance(self._session._conn, MockServerConnection):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe.py:3439]: Is it possible to not have this branch and mock the underlying analyzer?

File: src/snowflake/snowpark/mock/plan.py [hunk 2]
```diff
@@ -847,7 +849,11 @@ def calculate_expression(
         elif isinstance(exp, Pow):
             new_column = left**right
         elif isinstance(exp, EqualTo):
-            new_column = left == right
+            _left = left.reset_index(drop=True)
```
[the remark this answers is not in the document]
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/mock/plan.py:852]: Is overriding reset_index a good way to keep the sf_type?

File: src/snowflake/snowpark/mock/functions.py [hunk 3]
```diff
@@ -393,3 +393,8 @@ def mock_to_timestamp(column: ColumnEmulator, fmt: Union[ColumnEmulator, str] =
         sf_type=ColumnType(TimestampType(), column.sf_type.nullable),
         dtype=object,
     )
+
+
+@patch("row_number")
+def mock_row_number(window: TableEmulator, row_idx: int):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/functions.py:399]: Another PR is for windowing function. I'm curious if that is required to have the `row_number` work.

File: src/snowflake/snowpark/mock/plan.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/plan.py:852]: We could override `reset_index` for sure. Lemme poke around to see if there's a better way. https://pandas.pydata.org/docs/development/extending.html#subclassing-pandas-data-structures

**Comment by sfc-gh-stan** [comment]
> To be frank my biggest confusion is: We are enabling `DataFrame#DropDuplicates`, but we don't have any change in `drop_duplicates` function 🤔

This is because `DataFrame.drop_duplicates` is implemented using window functions https://github.com/snowflakedb/snowpark-python/blob/00ca832ce1727f1b6b969093ffc5d95812fc8c40/src/snowflake/snowpark/dataframe.py#L1575. This is stacked PR based on my other branch which implements window functions support.

File: src/snowflake/snowpark/mock/functions.py (hunk 3 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:399]: Yes, this is a stacked PR, I'm basing off this branch on [local/support-window-functions](https://github.com/snowflakedb/snowpark-python/tree/local/support-window-functions)

**Pull request closed without merging**